Prepare loading GHMHD initial data from files

nilsvu opened the pull request:
## Proposed changes

Top is the ID grid and bottom is the evolution grid:

![Bildschirm­foto 2023-03-15 um 15 36 10](https://user-images.githubusercontent.com/746230/225458802-2bf871e0-d3fc-40eb-9288-a80b521b75b8.png)


### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Move subcell actions to initialize_initial_data_dependent_quantities

Subcell initialization runs the TCI over the initial data to determine
if elements should switch to subcell already for the first timestep.
This is needed so non-smooth initial data like star surfaces can
be evolved.
Therefore, the subcell actions need to run after numeric initial data
has been loaded.

nilsdeppe commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/BoundaryConditions/Factory.hpp:
```diff
@@ -62,7 +62,6 @@ using standard_boundary_conditions = tmpl::push_back<
 
 /// Boundary conditions that work with finite difference.
 using standard_fd_boundary_conditions =
-    tmpl::list<ConstraintPreservingFreeOutflow, DirichletAnalytic,
```
Why do we need to change the definition of standard? Please keep as-is

nilsvu replied:
As mentioned earlier the DirichletAnalytic boundary condition doesn't work with numeric initial data. Since the `standard_fd_boundary_conditions` alias is only used in the executable and in tests I removed DirichletAnalytic here and appended it to the list where it's needed.

nilsdeppe commented on src/PointwiseFunctions/Hydro/Tags.hpp:
```diff
@@ -16,6 +16,19 @@
 namespace hydro {
+
+/// %Tags for options of hydrodynamic systems.
+namespace OptionTags {
```
Move these changes into a separate commit

nilsdeppe commented on src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp:
```diff
@@ -318,10 +318,25 @@ template <bool UseDgSubcell,
       EvolutionMetavarsDerived<InitialData, InterpolationTargetTags...>;
-  using defaults = GhValenciaDivCleanDefaults<UseDgSubcell>;
+
+  static constexpr bool use_dg_subcell = UseDgSubcell;
```
I don't understand why any of the changes in this commit are happening. This does more than the template parameter. Also, in a couple commits I ask if we can check locally if the analytic tag is retrievable, and if not then we assume numeric initial data. I think if that's possible then we can skip this commit completely.

nilsvu replied:
dropped this commit

nilsdeppe commented on src/Evolution/DgSubcell/Actions/Initialize.hpp:
```diff
@@ -198,17 +204,33 @@ struct Initialize {
-            make_not_null(&box), [&subcell_mesh](const auto prim_vars_ptr) {
-              prim_vars_ptr->initialize(subcell_mesh.number_of_grid_points());
-            });
+      if constexpr (UseNumericInitialData) {
```
Can you more easily check if the `AnalyticData` or `AnalyticSolution` tag is in the box? Then we don't need to add a template parameter to this action. I honestly don't love how invasive the numerical initial data is currently (not just in this PR). It causes an avalanche of template parameters to propagate through a lot of the code. It's almost as invasive as DG-FD, I find.

nilsvu replied:
done

nilsdeppe commented on src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp:
Can you specify in the commit message why this needs to be done, please?

nilsvu replied:
done

nilsdeppe reviewed (changes requested):
I'd like to deal with the last 2 commits in a separate PR. This is a rather hefty PR as-is

nilsvu committed:
Add EOS tag from options

To be used for numeric initial data where we can't get the EOS
from the analytic solution / data.

nilsvu committed:
Get GHMHD executable to compile again for numeric ID

nilsvu committed:
Project numeric initial data to subcell grids

nilsvu commented:
Made updates and shortened the PR.

nilsdeppe commented:
CI is very unhappy
